app/src/main/java/io/appium/uiautomator2/model/XPathFinder.java

```diff
@@ -51,235 +51,197 @@
+        final String uuid = UUID.randomUUID().toString();
+        UI_ELEMENTS_MAP.put(uuid, uiElement);
+
+        /*
+         * Setting the Element's className field.
+         * Reason for setting className field in Element object explicitly,
+         * className property might contain special characters like '$' if it is a Inner class and
+         * just not possible to create Element object with special characters.
+         * But Appium should consider Inner classes i.e special characters should be included.
+         */
+        setNodeLocalName(element, className);
+
+        setAttribute(element, Attribute.INDEX, String.valueOf(uiElement.getIndex()));
+        setAttribute(element, Attribute.CLASS, className);
+        setAttribute(element, Attribute.RESOURCE_ID, uiElement.getResourceId());
+        setAttribute(element, Attribute.PACKAGE, uiElement.getPackageName());
+        setAttribute(element, Attribute.CONTENT_DESC, uiElement.getContentDescription());
+        setAttribute(element, Attribute.TEXT, uiElement.getText());
+        setAttribute(element, Attribute.CHECKABLE, uiElement.isCheckable());
+        setAttribute(element, Attribute.CHECKED, uiElement.isChecked());
+        setAttribute(element, Attribute.CLICKABLE, uiElement.isClickable());
+        setAttribute(element, Attribute.ENABLED, uiElement.isEnabled());
+        setAttribute(element, Attribute.FOCUSABLE, uiElement.isFocusable());
+        setAttribute(element, Attribute.FOCUSED, uiElement.isFocused());
+        setAttribute(element, Attribute.SCROLLABLE, uiElement.isScrollable());
+        setAttribute(element, Attribute.LONG_CLICKABLE, uiElement.isLongClickable());
+        setAttribute(element, Attribute.PASSWORD, uiElement.isPassword());
+        if (uiElement.hasSelection()) {
+            element.setAttribute(Attribute.SELECTION_START.getName(),
+                    Integer.toString(uiElement.getSelectionStart()));
+            element.setAttribute(Attribute.SELECTION_END.getName(),
+                    Integer.toString(uiElement.getSelectionEnd()));
+        }
+        setAttribute(element, Attribute.SELECTED, uiElement.isSelected());
+        element.setAttribute(Attribute.BOUNDS.getName(),
+                uiElement.getBounds() == null ? null : uiElement.getBounds().toShortString());
+        element.setAttribute(UUID_ATTRIBUTE, uuid);
 
-  private static void setNodeLocalName(Element element, String className) {
-    try {
-      Field localName = element.getClass().getDeclaredField("localName");
-      localName.setAccessible(true);
-      localName.set(element, tag(className));
-    } catch (NoSuchFieldException e) {
-      Logger.error("Unable to set field localName:"+e.getMessage());
-    } catch (IllegalAccessException e) {
-      Logger.error("Unable to set field localName:"+e.getMessage());
+        for (UiElement<?, ?> child : uiElement.getChildren()) {
+            element.appendChild(toDOMElement(child, document));
```

Comment:
Could we get rid of the recursion here to avoid stack overflow exception?

Reply:
do you think it is going to be a problem?

Reply:
I'm afraid, that UiElement's interface does not allow to use non-recursive approach :(